### PR TITLE
STEP-2 Add Functionality to Add Task

### PR DESCRIPTION
**Pull request by sai-wtag**
## Description

[STEP-2](https://trello.com/c/2rCAjqWF/2-step-2-add-functionality-to-add-tasks)
This PR introduces the functionality to add a single task.

## Changes
1. Completing the task of adding into the the store,
2. Showing the list with minimal design,
3. Testing of few components.

## Review
Please have a review on this PR and comment if any changes required.
Screenshot 2023-01-23 at 5 44 04 PM
Screenshot 2023-01-23 at 5 44 33 PM

### CODE REVIEW

File: components/demo.vue [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  text-decoration: none;
+  font-size: 1.5rem;
+
+  &:hover {
+    text-decoration: underline;
+    color: $hovered-color;
+    cursor: pointer;
+  }
+}
+</style>
```
Comment (hhu-wtag) [inline components/demo.vue]: Is this file needed? If not we can delete it.

File: components/todo/AddTodoCard.vue [hunk 2]
```diff
@@ -0,0 +1,68 @@
+<template>
+  <div class="add-todo-card">
+    <form @submit.prevent="addTodo">
+      <div class="add-todo-card__body">
+        <!-- <textarea
+            v-model="form.title"
+            class="add-todo-card__input"
+            placeholder="Add Task"
+          /> -->
```
Comment (hhu-wtag) [inline components/todo/AddTodoCard.vue]: Please remove if not needed.

File: components/todo/Index.vue [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    <header class="header">
+      <h1>Todos</h1>
+      <SearchBar />
+    </header>
+    <TodoContainer />
+  </div>
+</template>
+<script>
+import SearchBar from './SearchBar.vue'
+import TodoContainer from './TodoContainer.vue'
```
Comment (hhu-wtag) [inline components/todo/Index.vue]: Please use relative path.

File: components/todo/TodoContainer.vue [hunk 4]
```diff
@@ -0,0 +1,75 @@
+<template>
+  <div class="todo-container">
+    <h2>Add Task</h2>
+    <div class="todo__header">
+      <button @click.prevent="$store.commit('todos/setCreate')">Create</button>
+      <FilterOptions />
+    </div>
+
+    <div class="card-container">
+      <!-- Add todo card -->
```
Comment (hhu-wtag) [inline components/todo/TodoContainer.vue]: We can remove this comment as the component name is self explanatory.

File: components/todo/TodoItem.vue [hunk 5]
```diff
@@ -0,0 +1,44 @@
+<template>
+  <div class="todo-item">
+    <div class="todo-title">{{ todo.title }}</div>
+    <div>
+      <div class="created-at">
+        Created at:
+        {{ $moment(todo.createdAt).format('MMMM Do YYYY') }}
```
Comment (hhu-wtag) [inline components/todo/TodoItem.vue]: `moment.js` is deprecated and isn't advised to use any more. [Link](https://momentjs.com/docs/#/-project-status/)
Please refer to more modern and future proof solution such as [date-fns](https://date-fns.org/)

File: nuxt.config.js [hunk 6]
```diff
@@ -34,6 +34,7 @@ export default {
     // https://go.nuxtjs.dev/axios
     '@nuxtjs/axios',
     '@nuxtjs/i18n',
+    '@nuxtjs/moment',
```
Comment (hhu-wtag) [inline nuxt.config.js]: ditto
tl;dr: ditto means something is said in the previous string of conversation and is mentioned here. e.g: A comment was made for moment.js and I'm mentioning it here.

File: test/LandingPage.test.js [hunk 7]
```diff
@@ -1,20 +1,11 @@
...
-      },
-      stubs: {
-        NuxtLink: RouterLinkStub,
-      },
-    })
-    const link = wrapper.findComponent(RouterLinkStub)
-    expect(link.exists()).toBe(true)
+  test('todo component exists', () => {
+    const wrapper = shallowMount(LandingPage)
+    // const todoApp = wrapper.findComponent(TodoIndexComponent)
```
Comment (hhu-wtag) [inline test/LandingPage.test.js]: ditto

File: components/demo.vue [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+export default {
+  name: 'LandingPage',
+  data() {
+    return {
+      
+    }
+  },
+  computed: {
+    todos() {
+      return this.$store?.state?.todos?.list || []
```
Comment (hhu-wtag) [inline components/demo.vue]: We can use [mapGetters](https://vuex.vuejs.org/guide/getters.html) here.

**Review by hhu-wtag — CHANGES_REQUESTED** [review]
Some changes required.
Other than that, good work. Thanks you. 🚀

**Comment by hhu-wtag** [comment]
> This PR is about ,
> 
> 1. Completing the task of adding into the the store,
> 2. Showing the list with minimal design,
> 3. Testing of few components.

1. Please add the ticket A/C here.
2. Add some screenshots/videos of the functionality.

File: components/demo.vue (hunk 1 above)
Comment (sai-wtag) [reply components/demo.vue]: Ok. vai, i'll delete it.

File: components/todo/AddTodoCard.vue (hunk 2 above)
Comment (sai-wtag) [reply components/todo/AddTodoCard.vue]: Ok, this is of no use now.

File: components/todo/Index.vue (hunk 3 above)
Comment (sai-wtag) [reply components/todo/Index.vue]: Ok, I'll add relative path to all the files, which are imported.

File: components/todo/TodoContainer.vue (hunk 4 above)
Comment (sai-wtag) [reply components/todo/TodoContainer.vue]: Noted. Thanks

File: components/todo/TodoItem.vue (hunk 5 above)
Comment (sai-wtag) [reply components/todo/TodoItem.vue]: Ok, vai, noted.

File: nuxt.config.js (hunk 6 above)
Comment (sai-wtag) [reply nuxt.config.js]: Ok, vai, got it.

File: test/LandingPage.test.js (hunk 7 above)
Comment (sai-wtag) [reply test/LandingPage.test.js]: Ok, vai.

File: components/demo.vue (hunk 8 above)
Comment (sai-wtag) [reply components/demo.vue]: Ok, bhai, Thanks.

File: pages/index.vue [hunk 9]
```diff
@@ -1,60 +1,13 @@
+<template>
+  <TodoApp />
+</template>
 <script>
-import { mapMutations } from 'vuex'
+import TodoApp from '../components/todo/Index.vue'
```
Comment (bia-wtag) [inline pages/index.vue]: We could use `@` here instead of `../` to keep the import consistant across components (easier for refactoring afterwards).

File: pages/index.vue [hunk 10]
```diff
@@ -1,60 +1,13 @@
...
-      
-    }
-  },
-  computed: {
-    todos() {
-      return this.$store?.state?.todos?.list || []
-    }
+  name: 'Todo',
+  components: {
+    TodoApp,
```
Comment (bia-wtag) [inline pages/index.vue:10]: nitpicking: As the component works more like a wrapper than an app itself, we could name it like that.

File: components/todo/AddTodoCard.vue [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+          type="text"
+          class="add-todo-card__input"
+          placeholder="Add Task"
+        />
+      </div>
+      <div class="add-todo-card__footer">
+        <button type="submit" class="add-todo-card__footer__btn">
+          Add Task
+        </button>
+        <button>Delete</button>
```
Comment (bia-wtag) [inline components/todo/AddTodoCard.vue]: Clicking this button creates an empty todo as the form is submitted without a `title`, please check attached screenshot.

File: components/todo/AddTodoCard.vue [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    addTodo() {
+      this.$emit('addTodo', this.form)
+      this.form.title = ''
+    },
+  },
+}
+</script>
+
+<style scoped lang="scss">
+$card-padding: 10px;
```
Comment (bia-wtag) [inline components/todo/AddTodoCard.vue:92]: note: `scoped` variables are not accessible elsewhere, so we can also use direct values.

File: pages/index.vue (hunk 9 above)
Comment (sai-wtag) [reply pages/index.vue]: Yes, bhai, noted, I'll use the relative path afterwards. Thanks

File: components/todo/AddTodoCard.vue [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+form {
+  display: flex;
+  flex-direction: column;
+  height: 100%;
+  gap: 5px;
+}
+.add-todo-card__footer {
+  display: flex;
+  align-items: center;
+  gap: 10px 10px;
```
Comment (bia-wtag) [inline components/todo/AddTodoCard.vue]: note: we could use above variable here.

File: components/todo/TodoContainer.vue [hunk 14]
```diff
@@ -0,0 +1,75 @@
+<template>
+  <div class="todo-container">
+    <h2>Add Task</h2>
+    <div class="todo__header">
+      <button @click.prevent="$store.commit('todos/setCreate')">Create</button>
```
Comment (bia-wtag) [inline components/todo/TodoContainer.vue]: while we do use `commit` directly outside of store, it a good practice to only trigger `actions` from components. also, setting local state based on store value might be unnecessary here as we are not creating todo from other components.

File: pages/index.vue (hunk 10 above)
Comment (sai-wtag) [reply pages/index.vue:10]: Noted, vai, I'll name it 'TodoWrapper' and I'll consider the naming conventions afterwards.

File: components/todo/AddTodoCard.vue (hunk 11 above)
Comment (sai-wtag) [reply components/todo/AddTodoCard.vue]: Ok, bhai. Thanks

File: components/demo.vue [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+      toggle: 'todos/toggle'
+    })
+  }
+
+}
+</script>
+<template>
+  <div>
+    <NuxtLink :to="switchLocalePath('en')" class="lang">English</NuxtLink>
+    <NuxtLink :to="switchLocalePath('fr')" class="lang">Français</NuxtLink>
```
Comment (bia-wtag) [inline components/demo.vue]: Can we change language without using different paths for the routes (i.e. `/fr` or `/en`)?

File: components/todo/AddTodoCard.vue (hunk 12 above)
Comment (sai-wtag) [reply components/todo/AddTodoCard.vue:92]: Ok, bhai, noted.

File: components/todo/AddTodoCard.vue (hunk 13 above)
Comment (sai-wtag) [reply components/todo/AddTodoCard.vue]: Noted.

File: components/todo/TodoContainer.vue (hunk 14 above)
Comment (sai-wtag) [reply components/todo/TodoContainer.vue]: Noted.

File: store/todos.js [hunk 16]
```diff
@@ -1,26 +1,41 @@
...
+      id: new Date().getTime(),
+      title: todo.title,
+      completed: false,
+      createdAt: new Date(),
+      completedAt: null,
+    }
+
+    state.list = [newTodo, ...state.list]
+
+    state.isCreating = false
```
Comment (bia-wtag) [inline store/todos.js]: Please keep mutations of state variables seperated. For example, we could move the above line to be changed in action using the mutation `setCreate` or something else.

Comment (sai-wtag) [reply store/todos.js]: Ok, Noted. Thanks

File: test/todo/FilterOptions.test.js [hunk 17]
```diff
@@ -0,0 +1,23 @@
+import { shallowMount } from '@vue/test-utils'
+import FilterOptions from '@/components/todo/FilterOptions.vue'
+
+describe('FilterOptions.vue', () => {
```
Comment (bia-wtag) [inline test/todo/FilterOptions.test.js:4]: These APIs (describe, test, expect) seems to be unavailable in the test file. Please update jest config along with projects so they're available.

File: nuxt.config.js [hunk 18]
```diff
@@ -34,6 +34,7 @@ export default {
...
   ],
   i18n: {
```
Comment (bia-wtag) [inline nuxt.config.js:34]: Can we move i18n to json files?

File: components/todo/AddTodoCard.vue (hunk 11 above)
Comment (bia-wtag) [reply components/todo/AddTodoCard.vue]: Screenshot 2023-01-23 at 10 56 07 AM

File: test/todo/FilterOptions.test.js (hunk 17 above)
Comment (bia-wtag) [reply test/todo/FilterOptions.test.js:4]: Screenshot 2023-01-23 at 11 11 02 AM

File: nuxt.config.js (hunk 18 above)
Comment (sai-wtag) [reply nuxt.config.js:34]: Thanks bhai, I'll definitely go through this.

File: components/todo/AddTodoCard.vue [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+      this.$store.dispatch('todos/deleteCurrentTask')
+    },
+
+    checkValidation() {
+      const title = this.form.title.trim()
+      const minTitleLength = this.validationRules.title.minLength
+      const maxTitleLength = this.validationRules.title.maxLength
+
+      if (!title) {
+        return 'Please enter a title'
```
Comment (hhu-wtag) [inline components/todo/AddTodoCard.vue]: Use `i18n` here.

File: components/todo/AddTodoCard.vue [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+      }
+
+      if (title.length < minTitleLength) {
+        return `Title should be more than ${minTitleLength} characters`
```
Comment (hhu-wtag) [inline components/todo/AddTodoCard.vue]: ditto

File: components/todo/AddTodoCard.vue [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+      }
+
+      if (title.length > maxTitleLength) {
+        return `Title should be less than ${maxTitleLength} characters`
```
Comment (hhu-wtag) [inline components/todo/AddTodoCard.vue]: ditto

**Review by hhu-wtag — APPROVED** [review]
Approved with some changes. Nice work. Thanks 🚀

File: test/todo/FilterOptions.test.js (hunk 17 above)
Comment (sai-wtag) [reply test/todo/FilterOptions.test.js:4]: Resolved

File: store/todos.js (hunk 16 above)
Comment (sai-wtag) [reply store/todos.js]: Resolved

**Review by bia-wtag — APPROVED** [review]
Tested locally. LGTM. Great work :clap:
